tests/components/icloud/test_config_flow.py

```diff
@@ -0,0 +1,284 @@
+"""Tests for the iCloud config flow."""
+import logging
+from unittest.mock import patch
+import pytest
+
+from pyicloud import PyiCloudService
+from pyicloud.exceptions import PyiCloudFailedLoginException
+
+from homeassistant import data_entry_flow
+from homeassistant.components.icloud import config_flow
+from homeassistant.components.icloud.config_flow import (
+    CONF_TRUSTED_DEVICE,
+    CONF_VERIFICATION_CODE,
+)
+from homeassistant.components.icloud.const import (
+    DOMAIN,
+    CONF_ACCOUNT_NAME,
+    CONF_GPS_ACCURACY_THRESHOLD,
+    CONF_MAX_INTERVAL,
+    DEFAULT_GPS_ACCURACY_THRESHOLD,
+    DEFAULT_MAX_INTERVAL,
+)
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+
+from tests.common import MockConfigEntry
+
+_LOGGER = logging.getLogger(__name__)
+
+USERNAME = "username@me.com"
+PASSWORD = "password"
+ACCOUNT_NAME = "Account name 1 2 3"
+ACCOUNT_NAME_FROM_USERNAME = None
+MAX_INTERVAL = 15
+GPS_ACCURACY_THRESHOLD = 250
+
+TRUSTED_DEVICES = [
+    {"deviceType": "SMS", "areaCode": "", "phoneNumber": "*******58", "deviceId": "1"}
+]
+
+
+@pytest.fixture(name="session")
+def mock_controller_session():
+    """Mock a successful session."""
+    with patch("pyicloud.base.PyiCloudSession"):
+        yield
+
+
+@pytest.fixture(name="service")
+def mock_controller_service():
+    """Mock a successful service."""
+    with patch("pyicloud.base.PyiCloudService"):
+        PyiCloudService.authenticate.return_value = None
```

Comment:
We don't want to use the real class here. We want to create a mock, `Mock` or `MagicMock` instance that we set the correct attributes on, and then we set that mock as return value of the patched class.

Reply:
You mean like this `mock_service.return_value.authenticate.return_value = None` ?
Exemple from your previous comment https://github.com/home-assistant/home-assistant/pull/26590#discussion_r330001700 that I followed on commit "Fix tests 13" https://github.com/home-assistant/home-assistant/pull/26590/commits/7bb2ca485677ebcaa5edcf597931e013f02b4141

But I couln't make it work as it was (`authenticate` didn't returned None and `requires_2fa` didn't returned True or False as expected for instance).

---

Also, can we pass the module code and only return the wanted value ?
I am telling this because during the test `pyicloud.authenticate()` (and many more), it is making I/O (web request + creating folder with the cookie) and I don't think it should be since we are asking to Apple to log with an unknown account everytime ... and it slows the tests.

Reply:
Back to `MagicMock`, here is what I got :

<img width="1602" alt="Capture d’écran 2019-11-15 à 01 33 20" src="https://user-images.githubusercontent.com/14821482/68907495-467f2980-0748-11ea-9a69-1b43ececeac4.png">
<img width="1608" alt="Capture d’écran 2019-11-15 à 01 33 02" src="https://user-images.githubusercontent.com/14821482/68907500-4b43dd80-0748-11ea-9ad9-e978574a3072.png">

`requires_2fa` shouldn't be a MagicMock instance but a boolean and `trusted_devices` should be a list of dict.
I couldn't mock those fields while trying multiple things. They are actually `@property` in the pyicloud lib.